python/ray/data/_internal/logical/rules/predicate_pushdown.py

```diff
@@ -0,0 +1,64 @@
+from ray.data._internal.logical.interfaces import (
+    LogicalOperator,
+    LogicalOperatorSupportsPredicatePushdown,
+    LogicalPlan,
+    Rule,
+)
+from ray.data._internal.logical.operators.map_operator import Filter
+
+
+class PredicatePushdown(Rule):
+    """Pushes down predicates across the graph.
+
+    This rule performs the following optimizations:
+    1. Combines chained Filter operators with compatible expressions
+    2. Pushes filter expressions down to operators that support predicate pushdown
+    """
+
+    def apply(self, plan: LogicalPlan) -> LogicalPlan:
+        """Apply predicate pushdown optimization to the logical plan."""
+        dag = plan.dag
+        new_dag = dag._apply_transform(self._try_fuse_filters)
+        new_dag = new_dag._apply_transform(self._try_push_down_predicate)
+        return LogicalPlan(new_dag, plan.context) if dag is not new_dag else plan
+
+    @classmethod
+    def _is_valid_filter_operator(self, op: LogicalOperator) -> bool:
+        return isinstance(op, Filter) and op.is_expression_based()
+
+    @classmethod
+    def _try_fuse_filters(cls, op: LogicalOperator) -> LogicalOperator:
+        """Fuse consecutive Filter operators with compatible expressions."""
+        if not cls._is_valid_filter_operator(op):
+            return op
+
+        input_op = op.input_dependencies[0]
+        if not cls._is_valid_filter_operator(input_op):
+            return op
+
+        # Combine predicates
+        combined_predicate = op._predicate_expr & input_op._predicate_expr
+
+        # Create new filter on the input of the lower filter
+        return Filter(
+            input_op.input_dependencies[0],
+            predicate_expr=combined_predicate,
+        )
+
+    @classmethod
+    def _try_push_down_predicate(cls, op: LogicalOperator) -> LogicalOperator:
+        """Push Filter down to any operator that supports predicate pushdown."""
+        if not cls._is_valid_filter_operator(op):
+            return op
+
+        input_op = op.input_dependencies[0]
+
+        # Check if the input operator supports predicate pushdown
+        if (
+            isinstance(input_op, LogicalOperatorSupportsPredicatePushdown)
+            and input_op.supports_predicate_pushdown()
+        ):
+            # Push the predicate down and return the result without the filter
+            return input_op.apply_predicate(op._predicate_expr)
+
+        return op
```

Comment:
You're only pushing the predicate into the read but not over other ops. I'm fine if you want to tackle it in a follow-up though